### PR TITLE
[Parse] Minor code improvement in function-type parsing

### PR DESCRIPTION
**Pull request by rintaro**
* Eliminate backtracking.
* Code cleanup.

In addition:

* Add fix-it for `rethrows` in function type.

### CODE REVIEW

File: lib/Parse/ParseType.cpp [hunk 1]
```diff
@@ -182,59 +183,46 @@ ParserResult<TypeRepr> Parser::parseType(Diag<> MessageID,
...
+  // Parse a throws specifier.
+  // Don't consume 'throws', if next token is not '->', so we can emit a more
+  // useful diagnostic when parsing a function decl.
   SourceLoc throwsLoc;
-  bool rethrows = false;
-  if (Tok.isAny(tok::kw_throws, tok::kw_rethrows) ||
-      (Tok.is(tok::kw_throw) && peekToken().is(tok::arrow))) {
-    if (Tok.is(tok::kw_throw)) {
+  if (Tok.isAny(tok::kw_throws, tok::kw_rethrows, tok::kw_throw) &&
+      peekToken().is(tok::arrow)) {
```
Comment (jrose-apple) [inline lib/Parse/ParseType.cpp]: This is definitely a behavior change for the recovery in a partial line:

```swift
let fn: () throws
```

Can you add tests for this case, and see if it's a net benefit?

Comment (rintaro) [inline lib/Parse/ParseType.cpp]: I'm not sure. What you mention is `Don't consume 'throws', if next token is not '->'`?
Swift 3.0.1 and this PR both emits:
```
test.swift:1:11: error: consecutive statements on a line must be separated by ';'
let fn: () throws
          ^
          ;
test.swift:1:12: error: expected expression
let fn: () throws
           ^
```
Actually, this behavior is from the original code:
https://github.com/apple/swift/pull/5721/files#diff-3d57a567ce2b696aa81ceb836f678fdeL222

**Comment by rintaro** [comment]
@jrose-apple  Anyway, I agree about bad recovery here.
But for now, I don't have good idea to fix that
(while keeping behavior for rdar://problem/20857518 https://github.com/apple/swift/commit/6951e460f036875687dfc8be00403928fd619db7)

Added the test case with `FIXME`.

File: include/swift/AST/DiagnosticsParse.def [hunk 2]
```diff
@@ -619,8 +619,9 @@ ERROR(expected_type_function_result,PointsToFirstBadToken,
       "expected type for function result", ())
 ERROR(generic_non_function,PointsToFirstBadToken,
       "only syntactic function types can be generic", ())
-ERROR(rethrowing_function_type,PointsToFirstBadToken,
-      "only function declarations may be marked 'rethrows'", ())
+ERROR(rethrowing_function_type,none,
+      "only function declarations may be marked 'rethrows'; "
+      "did you mean 'throws'?", ())
```
Comment (jtbandes) [inline include/swift/AST/DiagnosticsParse.def]: This doesn't make sense to me. The fact that `rethrows` isn't allowed in a function type (which I didn't realize until seeing this PR) seems like an unnecessary and surprising limitation. I would guess users will expect to be able to write `rethrows` function types, so suggesting `did you mean 'throws'?` will just be confusing.

Comment (jrose-apple) [inline include/swift/AST/DiagnosticsParse.def]: Summoning @rjmccall. A `rethrows` function type is useful, but it also makes the language more complicated, and I thought we decided not to do it.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@rintaro Can you please rebase?

**Comment by jrose-apple** [comment]
For the record, this was waiting on comments from @rjmccall about `rethrows`.

**Comment by slavapestov** [comment]
@jrose-apple I pinged John about this but haven't got a response yet. Since this patch is not changing the behavior of anything, can we go ahead and merge it?

**Comment by jrose-apple** [comment]
Fair point, sure.

**Comment by rjmccall** [comment]
That was definitely not a sneaky attempt to add rethrows function types to the language.  That would be a major addition.

**Comment by rintaro** [comment]
Sorry for delayed response, I was on vacation.
Thank you for merging this.

So, should I file a radar about ["GRAMMAR OF A FUNCTION TYPE" documentation issue](https://github.com/apple/swift/pull/5721#discussion_r89052442)?

**Comment by jrose-apple** [comment]
Seems like a good idea to me. It's not a critical issue, since the grammar generally allows more productions than the language, but it seems easy enough for the Apple DevPubs folks to just delete that line.

**Pull request merged**